db/compaction/compaction_job.cc

```diff
@@ -290,6 +294,141 @@ void CompactionJob::Prepare() {
   }
 }
 
+uint64_t CompactionJob::InitializeNumPlannedSubcompactions(
+    uint64_t max_output_files_limit, uint64_t ranges_size_limit) {
+  TEST_SYNC_POINT("CompactionJob::InitializeNumPlannedSubcompactions:0");
+  TEST_SYNC_POINT("CompactionJob::InitializeNumPlannedSubcompactions:1");
+  auto* c = compact_->compaction;
+  uint64_t max_subcompactions_limit = std::max(
+      std::uint64_t(1), static_cast<uint64_t>(c->max_subcompactions()));
+  uint64_t num_planned_subcompactions;
+  if (c->immutable_options()->compaction_pri == kRoundRobin) {
+    // For round-robin compaction prioity, we need to employ more
+    // subcompactions (may exceed the max_subcompaction limit). The extra
+    // subcompactions will be executed using reserved threads and taken into
+    // account bg_compaction_scheduled or bg_bottom_compaction_scheduled.
+
+    // Initialized against max_output_file_limit, the number of input
+    // files in the start level, and also the range size limit
+    num_planned_subcompactions = std::min(
+        {static_cast<uint64_t>(c->num_input_files(0)), max_output_files_limit,
+         static_cast<uint64_t>(ranges_size_limit)});
+
+    if (max_subcompactions_limit < num_planned_subcompactions) {
+      // Reserve resources when max_subcompaction is not sufficient
+      ReserveSubcompactionResources(
```

Comment:
I was originally thinking of calling `ReserveSubcompactionResources()` in the level of `GenSubcompactionBoundaries()` so that it's symmetric with `ReleaseSubcompactionResources()`. I should've made my suggestion more clear but I'm also good with the current writing style given we are prioritizing landing this.

Reply:
Does it help if we call this function `InitializeSubcompactionResources()` and then `ReleaseSubcompactionResources()` appears analogous (initialize raises `extra_num_subcompaction_threads_reserved_` from zero, while release returns it back to zero). Or if we call it `AcquireSubcompactionResources()` that'd be even clearer (assuming we agree that these really do operate at the same layer).